installer_show_maps_data/prep_data_for_maps.py

```diff
@@ -0,0 +1,315 @@
+    columns={
+        "local_authority_updated": "Local Authority",
+        "company_unique_id": "Total number of installation companies",
+    },
+    inplace=True,
+)
+
+# %%
+cumulative_installers
+
+# %%
+cumulative_installers.to_csv(
+    "s3://asf-exploration/installer_show_maps_data/outputs/cumulative_installers.csv"
+)
+
+# %%
+cumulative_installers[
+    "Total number of installation companies"
+].max(), cumulative_installers["Total number of installation companies"].min()
+
+# %% [markdown]
+# ## Installations and installers yearly
+
+# %% [markdown]
+# Number of installations per local authority per year:
+
+# %%
+installations_per_year = installations_data.groupby(
+    ["local_authority_updated", "commission_year"]
+).count()[["commission_date"]]
+installations_per_year.rename(
+    columns={"commission_date": "Number of installations"}, inplace=True
+)
+installations_per_year
+
+# %%
+installations_per_year = installations_per_year.unstack().fillna(0)
+installations_per_year = installations_per_year.droplevel(0, axis=1)
+installations_per_year["total"] = installations_per_year.sum(axis=1)
+installations_per_year
+
+# %% [markdown]
+# Number of installers per local authority per year:
+
+# %%
+installers_per_year = installations_data.groupby(
+    ["local_authority_updated", "commission_year"]
+).nunique()[["company_unique_id"]]
+installers_per_year.rename(
+    columns={"company_unique_id": "Number of installation companies per year"},
+    inplace=True,
+)
+installers_per_year
+
+# %%
+installers_per_year = installers_per_year.unstack().fillna(0)
+installers_per_year = installers_per_year.droplevel(0, axis=1)
+installers_per_year.reset_index(inplace=True)
+
+# %%
+installers_per_year = installers_per_year.merge(
+    cumulative_installers,
+    left_on="local_authority_updated",
+    right_on="Local Authority",
+    how="left",
+)
+
+# %%
+
+
+# %% [markdown]
+# ## Growth rate of installations and installers between 2021 and 2023
+
+# %% [markdown]
+# Should we not map when # installations in first year is 0? Or should we say the growth is 0?
+# I went with the first one, but happy to change it.
+
+# %%
+def growth_rate(final_year, first_year):
+    if first_year == 0:
+        return None
+    return (final_year - first_year) / first_year * 100
+
+
+# %%
+first_year = 2021
+final_year = 2023
+
+# %% [markdown]
+# Growth rate of installations:
+
+# %%
+installations_per_year[
+    f"growth_rate_{first_year}_{final_year}"
+] = installations_per_year.apply(
+    lambda x: growth_rate(x[final_year], x[first_year]), axis=1
+)
+
+# %%
+installations_per_year = installations_per_year.merge(
+    flourish_LAs, left_on="local_authority_updated", right_on="Name", how="left"
+)
+
+# %%
+installations_per_year["country"] = installations_per_year["Code"].str[0]
+installations_per_year["country"] = installations_per_year["country"].map(
+    {"E": "England", "W": "Wales", "S": "Scotland", "N": "Northern Ireland"}
+)
```

Comment:
fyi for some reason `installations_per_year.iloc[7]` does not have local authority code data and therefore no country assigned.

Reply:
yeah... that happens! we're not plotting those.

Reply:
yup, these are installations with no location information. Nothing we can do about it unfortunately. Those don't get mapped.